relayer/chain/ethereum/message.go

```diff
@@ -150,12 +150,49 @@ func MakeMessageFromEvent(event *etypes.Log, receiptsTrie *etrie.Trie, log *logr
 
 	value := hex.EncodeToString(message.Data)
 	log.WithFields(logrus.Fields{
-		"payload":     value,
-		"blockNumber": message.VerificationInput.AsBasic.BlockNumber,
-		"eventIndex":  message.VerificationInput.AsBasic.EventIndex,
+		"payload":    value,
+		"blockHash":  message.VerificationInput.AsReceiptProof.BlockHash.Hex(),
+		"eventIndex": message.VerificationInput.AsReceiptProof.TxIndex,
 	}).Debug("Generated message from Ethereum log")
 
-	msg := chain.Message{AppID: event.Address, Payload: message}
+	return &message, nil
+}
+
+func MakeMessageChunker(messagesByAddress map[common.Address][]*Message, chunkSize int) func() ([]chain.Message, bool) {
+	i, offset := 0, 0
+	addresses := make([]common.Address, 0, len(messagesByAddress))
+	for k := range messagesByAddress {
+		addresses = append(addresses, k)
+	}
 
-	return &msg, nil
+	return func() ([]chain.Message, bool) {
+		chunk := make([]chain.Message, 0)
+		count := 0
+
+		for i < len(addresses) {
+			if count == chunkSize {
+				return chunk, true
+			}
```

Comment:
If `chunkSize == 0` (which might not make much sense, but is a valid input), this leads to an infinite loop in the caller.